source/llassetgen/source/Packing.cpp

```diff
@@ -16,43 +26,33 @@ namespace llassetgen {
             }
 
             if (usedHeight + rectSize.y > packing.atlasSize.y) {
-                return false;
+                if (allowGrowth) {
+                    PackingSizeType finalHeight = usedHeight + rectSize.y;
+                    auto numDoublings = ceilLog2(ceilDiv(finalHeight, packing.atlasSize.y));
+                    packing.atlasSize.y <<= numDoublings;
+                } else {
+                    return false;
+                }
             }
 
             store(rectSize);
             return true;
         }
 
-        bool ShelfNextFitPacker::packNextRotatable(Vec2<PackingSizeType> rectSize) {
+        bool ShelfNextFitPacker::packNextWithRotations(Vec2<PackingSizeType> rectSize) {
             PackingSizeType minSide, maxSide;
             std::tie(minSide, maxSide) = std::minmax(rectSize.x, rectSize.y);
             PackingSizeType remainingWidth = packing.atlasSize.x - currentShelfSize.x;
             PackingSizeType remainingHeight = packing.atlasSize.y - usedHeight;
 
-            bool fitsUpright = currentShelfSize.y >= maxSide && remainingWidth >= minSide;
-            bool fitsSideways = remainingWidth >= maxSide && remainingHeight >= minSide;
-            bool upright = fitsUpright;
-            if (!fitsUpright && !fitsSideways) {
-                openNewShelf();
-                if (usedHeight + minSide > packing.atlasSize.y) {
-                    return false;
-                }
-
-                if (maxSide > packing.atlasSize.x) {
-                    if (usedHeight + maxSide > packing.atlasSize.x) {
-                        return false;
-                    }
-
-                    upright = true;
-                } else {
-                    upright = false;
-                }
-            }
-
-            if (upright) {
+            if (currentShelfSize.y >= maxSide && remainingWidth >= minSide) {
                 store({minSide, maxSide});
-            } else {
+            } else if (remainingWidth >= maxSide && remainingHeight >= minSide) {
                 store({maxSide, minSide});
+            } else {
+                openNewShelf();
+                return maxSide > packing.atlasSize.x ? packNextNoRotations({minSide, maxSide})
+                                                     : packNextNoRotations({maxSide, minSide});
```

Comment:
This is in order to let `packNextNoRotations` do the growing if needed, right? Seems a bit convoluted, maybe you could extract the growing into a separate function?

Reply:
Yeah, in hindsight that was not very clear. I did what you suggested and factored the growth logic out into another method.